### ISSUE TITLE
vmselect does not serve partial data

### ISSUE THREAD

**Issue by istolpov** [VictoriaMetrics/VictoriaMetrics#1148]
# Premise

Hello. Today I found that when switching off one `vmstorage` node I start to get an error:
```
{
    "error": "error when exporting data for query=\"field1\" on the time range (start=1616675055000, end=1616682255000): cannot fetch data for \"AccountID=0, ProjectID=0, MinTimestamp=2021-03-25 12:24:15 +0000 UTC, MaxTimestamp=2021-03-25 14:24:15 +0000 UTC, TagFilters=[\\n{Key=\\\"\\\", Value=\\\"field1\\\", IsNegative: false, IsRegexp: false}\\n]\": error occured during search: cannot fetch query results from vmstorage nodes: cannot perform search on vmstorage 10.216.0.219:8401: cannot obtain connection from a pool: cannot dial 10.216.0.219:8401: dial tcp4 10.216.0.219:8401: connect: no route to host",
    "errorType": "422",
    "status": "error"
}
```

To prevent common answers:

1) No `-search.denyPartialResponse` as you can see.
```
      [root@vm00802 ~]# docker inspect vmselect | grep Cmd -A 3
                  "Cmd": [
                      "-storageNode=10.216.0.219:8401",
                      "-storageNode=vmstorage:8401"
                  ],
``` 

2) This is a test cluster so no "other node probably failed due to overload" shenanigans. There is no load. Here it is, running happily:
```
[root@vm00802 ~]# docker ps | grep vmstorage
849a28fb7435        registry*******/vmetrics_storage:v1.54.1-cluster   "/vmstorage-prod -..."   2 days ago          Up 21 minutes       0.0.0.0:8400-8401->8400-8401/tcp, 127.0.0.1:8482->8482/tcp   vmstorage
```

# Investigation

I have found the bug that basically **disables partial data serving on /api/v1/query.**

1) Here's where the outermost error comes from:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/a3b3d434a37ce5cb81d5954860386eac188ca62f/app/vmselect/prometheus/prometheus.go#L1042-L1044

2) prometheus.exportHandler's signature is the following:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/a3b3d434a37ce5cb81d5954860386eac188ca62f/app/vmselect/prometheus/prometheus.go#L301-L302 
  Take a note that `reduceMemUsage bool` is **set explicitly to `false`** (as opposed to the `prometheus.ExportHandler` where it is set from `reduce_mem_usage` request parameter).

3) In the exportHandler's body: 
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/a3b3d434a37ce5cb81d5954860386eac188ca62f/app/vmselect/prometheus/prometheus.go#L367-L372
And here we see the `cannot fetch data for` line that is in the original error at the very top of this issue.

# tl;dr
1) When `reduce_mem_usage == false` (default behavior!) VMetrics does not serve partial data via `/api/v1/export`
2) VMetrics does not serve partial data **at any circumstances** via `/api/v1/query` due to `reduce_mem_usage` hardcoded.

# Questions
1) Why `reduce_mem_usage` drives the logic behind partial data serving and why it is not documented anywhere.
2) Why `reduce_mem_usage` is hardcoded to false in case of `/api/v1/query`

# Notes
1) This was found in v1.54.1, however, `cluster` branch contains those lines of code so it is safe to assume it is also affected.

**Comment by istolpov** [comment]
This is probably related:
https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1086

**Comment by valyala** [comment]
Partial responses are denied unconditionally for [/api/v1/export* handlers](https://victoriametrics.github.io/#how-to-export-time-series), since users expect that these handlers return all the raw samples from the database. This cannot be satisfied when some `vmstorage` nodes are temporarily unavailable.

The `/api/v1/query` handler should return partial responses by default when it is used for querying calculated data. It unconditionally denies partial responses for e.g. `range queries`, which return raw samples from the database (aka data export). See https://valyala.medium.com/analyzing-prometheus-data-with-external-tools-5f3e5e147639 . Such queries aren't used neither by Grafana, nor by [vmalert](https://victoriametrics.github.io/vmalert.html), so this shouldn't prevent from serving partial responses for common case.

**Comment by istolpov** [comment]
Hi @valyala and thanks for the fast reply!

I understand the logic behind assuming `api/v1/export` requiring full data.
However, I don't understand the logic behind assuming user only wants full data by querying a range.
This is IMO a bold assumption -- some users wish to prioritize availability over consistency and when not "exporting" data explicitly they probably just want the data.

Maybe we could just add an error message stating exactly this -- that time ranges are available when replication constraints are satisfied.

Correct me if I'm wrong, but there is also "isPartial" flag in the response clearly indicating whether the data returned in the range is partial or not.

In any case: this must be clearly documented somewhere, because this makes the partial availability guarantee seem either like it's bugged or misrepresented in the documentation promises, akin to a small print, but in this case somebody forgot to print this "small print": https://victoriametrics.github.io/Cluster-VictoriaMetrics.html#cluster-availability

P.s. Forgive my lack of knowledge but how exactly does Grafana get the range of data to render if not by using range queries? If you have any article or something like that it would be helpful. I did not find any info on the topic.

**Comment by valyala** [comment]
> there is also "isPartial" flag in the response clearly indicating whether the data returned in the range is partial or not.

Yes. But the response data format for [/api/v1/export* handlers](https://victoriametrics.github.io/#how-to-export-time-series) handler has no place for inserting the `isPartial` flag.

> how exactly does Grafana get the range of data to render if not by using range queries?

Grafana sends queries to [/api/v1/query_range](https://prometheus.io/docs/prometheus/latest/querying/api/#range-queries). This handler always returns calculated samples on the `start ... end` time range with a constant `step` interval between samples. This handler should return partial responses if some vmstorage nodes are unavailable.

**Comment by istolpov** [comment]
Hi, @valyala!

What is the plan regarding the last message, particularly the `/api/v1/query` endpoint?

**Comment by valyala** [comment]
The `/api/v1/query` endpoint returns raw samples if the provided query contains only [series selector](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors) followed by a time range in square brackets (aka `range vector selector`) as described in [this article](https://valyala.medium.com/analyzing-prometheus-data-with-external-tools-5f3e5e147639). This means that it works as data export, so it follows the same logic as `/api/v1/export*` handlers - e.g. it doesn't return partial responses.

**Comment by istolpov** [comment]
@valyala could you, please, move/copy this info from your personal blog to the vmetrics documentation? The behavior is confusing.

Also, I'm still not sure why reduce_mem_usage variable drives the logic behind logic we are talking about.

**Comment by valyala** [comment]
The `exportHandler()` function always returns an error if at least a single vmstorage node is unavailable during querying:

- if `reduceMemUsage=true`, then `netstorage.ProcessSearchQuery()` is called with `denyPartialResponse=true`
- if `reduceMemUsage=false`, then `netstorage.ExportBlocks()` is called, which, in turn, calls `processSearchQuery()` with `denyPartialResponse=true`

**Comment by valyala** [comment]
> @valyala could you, please, move/copy this info from your personal blog to the vmetrics documentation? The behavior is confusing.

Done in the commit 0511d5c3a045f15eb54b84e5ded7ec6085d83036 .

Closing the issue as addressed then.
